src/ert/ensemble_evaluator/evaluator_async.py

```diff
@@ -0,0 +1,479 @@
+
+        # batching section
+        self._processing_queue: asyncio.Queue[
+            List[Tuple[Callable[[List[CloudEvent]], Awaitable[None]], CloudEvent]]
+        ] = asyncio.Queue()
+        self._max_batch_size: int = 500
+        self._batching_interval: int = 2
+
+    async def _publisher(self) -> None:
+        try:
+            while True:
+                msg = await self._messages.get()
+                await asyncio.gather(
+                    *[client.send(msg) for client in self._clients],
+                    return_exceptions=True,
+                )
+        except asyncio.CancelledError:
+            # when cancelling the task, make sure to send the rest
+            while not self._messages.empty():
+                msg = self._messages.get_nowait()
+                await asyncio.gather(
+                    *[client.send(msg) for client in self._clients],
+                    return_exceptions=True,
+                )
+
+    async def _append_message(self, snapshot_update_event: PartialSnapshot) -> None:
+        message = self._create_cloud_message(
+            EVTYPE_EE_SNAPSHOT_UPDATE,
+            snapshot_update_event.to_dict(),
+        )
+        if message:
+            await self._messages.put(message)
+
+    async def _process_buffer(self) -> None:
+        while True:
+            batch = await self._processing_queue.get()
+            function_to_events_map: Dict[
+                Callable[[List[CloudEvent]], Awaitable[None]], List[CloudEvent]
+            ] = OrderedDict()
+            for func, event in batch:
+                if func not in function_to_events_map:
+                    function_to_events_map[func] = []
+                function_to_events_map[func].append(event)
+
+            for func, events in function_to_events_map.items():
+                await func(events)
+
+            self._processing_queue.task_done()
+
+    async def _dispatcher(self) -> None:
+        event_handler = {}
+
+        def set_event_handler(event_types: Set[str], function: Any) -> None:
+            for event_type in event_types:
+                event_handler[event_type] = function
+
+        for e_type, f in (
+            (EVGROUP_FM_ALL, self._fm_handler),
+            ({EVTYPE_ENSEMBLE_STARTED}, self._started_handler),
+            ({EVTYPE_ENSEMBLE_STOPPED}, self._stopped_handler),
+            ({EVTYPE_ENSEMBLE_CANCELLED}, self._cancelled_handler),
+            ({EVTYPE_ENSEMBLE_FAILED}, self._failed_handler),
+        ):
+            set_event_handler(e_type, f)
+
+        while True:
+            batch: List[
+                Tuple[Callable[[List[CloudEvent]], Awaitable[None]], CloudEvent]
+            ] = []
+            start_time = asyncio.get_running_loop().time()
+            while (
+                len(batch) < self._max_batch_size
+                and asyncio.get_running_loop().time() - start_time
+                < self._batching_interval
+            ):
+                try:
```

Comment:
I'm a bit confused by this: `handle_dispatch` reads messages from a websocket, which are turned into events put into an async queue. Then the `_dispatcher` task gets those events from that queue and parses them into (event function, args) pairs before putting it into processing queues. I think it makes sense somewhat because it handles websocket->queue->processing buffer, but the naming of the functions or something like that could maybe somehow make this more clear

Reply:
> I'm a bit confused by this: `handle_dispatch` reads messages from a websocket, which are turned into events put into an async queue. Then the `_dispatcher` task gets those events from that queue and parses them into (event function, args) pairs before putting it into processing queues. I think it makes sense somewhat because it handles websocket->queue->processing buffer, but the naming of the functions or something like that could maybe somehow make this more clear

That's essentially the async version of batch processing, wherein nothing is blocking processing tasks and all exists in their own queues. Will try to improve the naming though.